crates/walrus-indexer/src/storage.rs

```diff
@@ -0,0 +1,471 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#![allow(dead_code)]
+
+//! Storage layer for the Walrus Indexer.
+//!
+//! This module implements the Walrus Index schema for indexing blobs and quilt patches.
+//! Indices are stored in RocksDB, organized by 'blob_manager' that provide isolated namespaces
+//! similar to S3 buckets.
+
+use std::{path::Path, sync::Arc};
+
+use anyhow::Result;
+use bcs;
+use rocksdb::{OptimisticTransactionDB, Options as RocksDbOptions, Transaction};
+use serde::{Deserialize, Serialize};
+use sui_types::base_types::ObjectID;
+use typed_store::{
+    TypedStoreError,
+    rocks::{DBMap, MetricConf},
+};
+use walrus_core::{
+    BlobId,
+    QuiltPatchId,
+    encoding::quilt_encoding::{QuiltIndexApi, QuiltPatchApi, QuiltPatchInternalIdApi},
+    metadata::QuiltIndex,
+};
+
+/// Primary index, blob_manager/identifier/sequence_number -> blob or quilt patch.
+const CF_NAME_PRIMARY_INDEX: &str = "walrus_index_primary";
+/// Object index, object_id -> blob_manager/identifier/sequence_number.
+const CF_NAME_OBJECT_INDEX: &str = "walrus_index_object";
+/// Sequence store, sequence_number.
+const CF_NAME_SEQUENCE_STORE: &str = "walrus_sequence_store";
+/// Quilt patch index, blob_id of a quilt patch -> quilt_patch_id.
+const CF_NAME_QUILT_PATCH_INDEX: &str = "walrus_quilt_patch";
+/// Sequence store, sequence_number.
+const SEQUENCE_KEY: &[u8] = b"sequence";
+
+/// Key structure for primary index: blob_manager/identifier/sequence_number.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct PrimaryIndexKey {
+    /// The blob manager ID.
+    blob_manager: ObjectID,
+    /// The identifier.
+    identifier: String,
+    /// The sequence number.
+    sequence_number: u64,
```

Comment:
question: what is sequence number? IIUC, blob_manager is the namespace, identifier is the user defined file name, is that correct?